Remove socket from sockmap after XCloseSocket call

nukeulater opened the pull request:
Don't play default sounds if custom sounds are not present.
Added in-game selection support for 4k, 1440p and 480p. (will probably not fix the text dissapearing  for high def resolutions)
Some code clean-up.

**IT CAN BE MERGED**

 Fixed clients not being able to join dedicated servers hosted on the same computer, but the other peers are getting kicked out if local clients are in.
But still, the changes are valid.

nukeulater committed:
Remove socket from sockmap after XCloseSocket call

Don't play default sounds if custom sounds are not present.

nukeulater committed:
Use safe version of _wfopen

nukeulater committed:
Merge remote-tracking branch 'origin/development' into misc

nukeulater committed:
Fixed clients not being able to join dedicated servers hosted on the same computer

Partially fixed CUser memory leak, by reusing memory already allocated by updating information in the structure.

num0005 commented on xlive/H2MOD/Modules/Config/Config.cpp:
```diff
@@ -171,7 +171,8 @@ void SaveH2Config() {
 	addDebugText(fileConfigPathLog);
-	FILE* fileConfig = _wfopen(fileConfigPath, L"wb");
+	FILE* fileConfig; 
+	_wfopen_s(&fileConfig, fileConfigPath, L"wb");
```
Check the return value and I'm not sure if just switching it like that is ok, _wfopen_s blocks any other program from accessing the file afaik.

nukeulater replied:
The files get closed after read/write, blocking shouldn't be a problem.

num0005 commented on xlive/H2MOD/Modules/Config/Config.cpp:
```diff
@@ -1855,7 +1856,7 @@ void ReadH2Config() {
 		swprintf(fileConfigPathLog, 1124, L"Reading Flag Config: \"%ws\"", fileConfigPath);
 		addDebugText(fileConfigPathLog);
-		fileConfig = _wfopen(fileConfigPath, L"rb");
+		_wfopen_s(&fileConfig, fileConfigPath, L"rb");
```
also here

num0005 commented on xlive/H2MOD/Modules/Config/Config.cpp:
```diff
@@ -1866,7 +1867,7 @@ void ReadH2Config() {
 			swprintf(fileConfigPathLog, 1124, L"Reading Config: \"%ws\"", fileConfigPath);
 			addDebugText(fileConfigPathLog);
-			fileConfig = _wfopen(fileConfigPath, L"rb");
+			_wfopen_s(&fileConfig, fileConfigPath, L"rb");
```
and here

num0005 commented on xlive/XLive/Networking/XSocket.cpp:
```diff
@@ -19,6 +21,10 @@ int WINAPI XWSACleanup()  	// XWSACleanup
 {
 	LOG_TRACE_XLIVE("XSocketClose: socket: {}", s);
+
+	if (userManager.sockmap[s])
```
that will insert an element, use `count` or `find`

num0005 commented on xlive/XLive/Networking/xlive_network.cpp:
```diff
@@ -266,21 +265,12 @@ int WINAPI XSocketRecvFrom(SOCKET s, char *buf, int len, int flags, sockaddr *fr
-				}
-				else // only create the user if they're not already in the system.
-					userManager.CreateUser((XNADDR*)(buf + 8), TRUE);
+				XNADDR* userxn = reinterpret_cast<XNADDR*>(buf + 8);
```
not your fault but would it be possible to use a struct for buf? It seems we know what's at every offset.

num0005 commented on xlive/xliveless.cpp:
```diff
@@ -187,7 +187,7 @@ BOOL SetDlcBasepath( int num )
 					// check file exists
 					FILE *fp_dlc;
-					fp_dlc = _wfopen( wnum.c_str(), L"rb" );
+					_wfopen_s(&fp_dlc, wnum.c_str(), L"rb" );
```
this file also has the issue with not checking the return value.

num0005 commented on xlive/XLive/UserManagement/CUser.cpp:
```diff
@@ -51,13 +49,29 @@ void CUserManagement::CreateUser(const XNADDR* pxna, BOOL user)
+		LOG_TRACE_NETWORK("CUserManagement::CreateUser() new secure address {:x}", pxna->inaOnline.s_addr);
+
+		nUser = new CUser;
+		memset(&nUser->xnaddr, 0x00, sizeof(XNADDR));
```
why are you clearing the addr if you are going to memcpy over it?

num0005 commented on xlive/XLive/UserManagement/CUser.cpp:
```diff
@@ -98,51 +112,41 @@ void CUserManagement::CreateUser(const XNADDR* pxna, BOOL user)
-			smap1 = std::make_pair(H2Config_ip_lan, ntohs(htons(to_remove_user->xnaddr.wPortOnline) + 1));
-		}
-		else
+		if (xnmap[ina.s_addr])
```
this also could insert unneeded elements into the array, use `count` or `find`, etc

num0005 commented on xlive/XLive/UserManagement/CUser.cpp:
```diff
@@ -98,51 +112,41 @@ void CUserManagement::CreateUser(const XNADDR* pxna, BOOL user)
-			smap = std::make_pair(to_remove_user->xnaddr.ina.s_addr, to_remove_user->xnaddr.wPortOnline);
-			smap1 = std::make_pair(to_remove_user->xnaddr.ina.s_addr, ntohs(htons(to_remove_user->xnaddr.wPortOnline) + 1));
+			if (it.second == ina.s_addr) {
+				secure_map.erase(it.first);
```
not sure if it's safe to remove elements while iterating? Could crash or skip an element... Might be better to use std::remove_if

nukeulater replied:
it exits the loop after it finds the element, shouldn't cause any issues

num0005 replied:
ok yeah that's fine then.

shayaantx replied:
is it fine though? how many different threads might access this map at the same time? if we can guarantee single threaded access to the map then it seems okay, but in reality it should be thread safe

nukeulater committed:
Fixed unordered_map element erasing (big oops)

Created struct for secure packet.

nukeulater committed:
Fixed wrong function definitions for entity packet encode/decode

nukeulater committed:
Removed duplicate code from map load function.

nukeulater committed:
Check return value of _wfopen_s

num0005 commented on xlive/XLive/Networking/xlive_network.cpp:
```diff
@@ -262,25 +261,17 @@ int WINAPI XSocketRecvFrom(SOCKET s, char *buf, int len, int flags, sockaddr *fr
 		if (iplong != H2Config_master_ip)
 		{
-			if (*(ULONG*)buf == annoyance_factor && ret == 12 + sizeof(XNADDR))
+			if (ret == sizeof(SecurePacket)
```
would be nicer to call it something other than `ret`, maybe `packet_size`. Could also merge the two if statements.

nukeulater replied:
the if statements can't be merged

num0005 replied:
I missed the rest of the code that handles packets that aren't SecurePacket

num0005 commented on xlive/XLive/UserManagement/CUser.cpp:
```diff
@@ -204,13 +209,12 @@ void CUserManagement::ConfigureLocalUser(XNADDR* pxna, ULONGLONG xuid, char* use
 	local_user.secure.s_addr = pxna->inaOnline.s_addr;
-	this->secure_packet = new char[12 + sizeof(XNADDR)];
-	memset(secure_packet, NULL, 12 + sizeof(XNADDR));
+	memset(&securePacket, NULL, sizeof(SecurePacket));
```
using `0` instead of `NULL` or the `ZeroMemory` macro would make it more clear. `NULL` could make it seem like you are dealing with an array of pointers.

num0005 commented on xlive/xliveless.cpp:
```diff
@@ -972,8 +972,7 @@ DWORD WINAPI XStorageUploadFromMemory( DWORD dwUserIndex, const WCHAR *wszServer
 	FILE *fp;
-
-	fp = _wfopen( wszServerPath, L"wb" );
+	_wfopen_s(&fp, wszServerPath, L"wb" );
```
still need to check return value here, per MSDN

num0005 commented:
You might want to consider using an enum for GametypesMap, so other code doesn't need to know the exact string it checks against.

NukeULater commented:
Not sure I understand.

nukeulater committed:
Code review changes

nukeulater committed:
Cleanup

nukeulater committed:
Added custom resolutions support within the game

nukeulater committed:
Replaced nulling buffer memset with SecureZeroMemory

Misc code cleanup.
Added 4k and 1440p in-game resolution selection support.
Updated some stuff in the custom packets code and serverlist.

shayaantx commented on xlive/XLive/UserManagement/CUser.cpp:
```diff
@@ -12,15 +10,17 @@
+SecurePacket securePacket;
 CUserManagement userManager;
-const DWORD annoyance_factor = 0x11223344;
+const DWORD annoyance_factor = 0x8E0A40F1;
```
uhhh, please remove

shayaantx commented on xlive/H2MOD/Modules/Networking/CustomPackets/CustomPackets.cpp:
```diff
@@ -162,13 +164,13 @@ void __stdcall message_gateway_hook_2(void *thisx, int a2, int message_type, int
 	}
-	case e_network_message_types::map_file_name:
+	case map_file_name:
 		return;
```
so for map_file_name we don't want to call p_network_message_gateway_2 but for everything other type we do? 

there should be a single line doc explaining this and why

nukeulater committed:
Oops

shayaantx commented on xlive/H2MOD.cpp:
```diff
@@ -293,7 +292,7 @@ signed int __cdecl object_new_hook(void *pObject)
 	object_to_variant[result] = variant_index;
 
 	//wchar_t DebugText[255] = { 0 };
-	//ZeroMemory(DebugText, sizeof(DebugText));
```
please revert back to ZeroMemory unless you have a real reason to secure this data

shayaantx commented on xlive/H2MOD.cpp:
```diff
@@ -881,7 +877,7 @@ void H2MOD::CustomSoundPlay(const wchar_t* soundName, int delay)
 			std::this_thread::sleep_until(timePoint);
 
-		PlaySound(soundName, NULL, SND_FILENAME);
+		PlaySound(soundName, NULL, SND_FILENAME | SND_NODEFAULT);
```
when is soundName null/empty? why even go through any of this code if a null soundName was passed in? check add a null check and return

shayaantx commented on xlive/H2MOD.cpp:
```diff
@@ -36,23 +39,19 @@ bool b_Infection = false;
-extern XUID xFakeXuid[4];
-std::unordered_map<int, int> object_to_variant;
-
+std::unordered_map<wchar_t*, bool&> GametypesMap
```
In reference to https://github.com/PermaNulled/cartographer/pull/242#issuecomment-514374292 hes probably right, an enum would be cleaner, however...

What you want to stride for is not having a custom map, you should be thinking, if 2moro someone wanted to add a new game type in code how many different spots would they need to update.

It'd be very nice if this map was driven by all the classes that implement GameClientServerHandler (i.e., register each sub class of GameClientServerHandler in some map). Then embed the string constant and whether its enabled in those classes as well. That way adding a new game type just involves implementing GameClientServerHandler. This would be the polymorphic way of handling this

You can even improve a lot of the event code (like player spawn and player death), for example

Right now there is something like below

```
if (b_HeadHunter)
	{
		DatumIndex dead_player = DatumIndex(unit_datum_index);
		headHunterHandler->playerDeath->SetDeadPlayer(dead_player); // set this so we can spawn a skull on their position.
		headHunterHandler->playerDeath->execute();
	}
	if (b_GunGame) {
		gunGame->playerDeath->setUnitDatumIndex(unit_datum_index);
		gunGame->playerDeath->execute();
	}
```

but it really should be

```	
for (auto handler : game_handlers) {
		if (handler.enabled()) {
			handler->playerDeath.SetDeadPlayer(dead_player);
			handler->playerDeath->execute();
		}
	}
```

You can get rid of a lot of duplicated code like this
